### PR TITLE
Stork Health monitor improvements

### PR DESCRIPTION
**Pull request by diptiranjanpx**
**What type of PR is this?**
improvement

**What this PR does / why we need it**:
1. Filtering pods wrt node instead of getting all pods to work on and looping on those to filter out the pods running in offline node.
2. Batch delete of pods in a batch of 5.
3. Do a fresh check of node status before issuing batch delete.
4. Indexing spec.nodeName in informercache to filter out pods running on a node.

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
TBD
```release-note
Issue:
User Impact:
Resolution

```

**Does this change need to be cherry-picked to a release branch?**:
no

**Test**:
UTs - 2 tests from current UT are failing as these need proper mocking of driver.InspectNode() . Will push the fix.

Manual test : 

pods running a node 
```
➜  stork git:(PWX-38468) ✗ kubectl get pods --all-namespaces --field-selector spec.nodeName=ip-10-13-193-24.pwx.purestorage.com
NAMESPACE     NAME                                              READY   STATUS    RESTARTS       AGE
busy-10       busybox-deployment-6dcd7756f9-7q5r6               1/1     Running   0              25m
busy-14       busybox-deployment-6dcd7756f9-ghz9f               1/1     Running   0              25m
busy-16       busybox-deployment-6dcd7756f9-sqdnm               1/1     Running   0              41m
busy-2        busybox-deployment-6dcd7756f9-w7v7r               1/1     Running   0              25m
busy-20       busybox-deployment-654cf6f86f-zf9r2               1/1     Running   0              25m
busy-4        busybox-deployment-6dcd7756f9-zqrm7               1/1     Running   0              25m
busy-6        busybox-deployment-6dcd7756f9-z2nnz               1/1     Running   0              41m
busy-7        busybox-deployment-6dcd7756f9-bkwt9               1/1     Running   0              25m
es            elasticsearch-data-0                              1/1     Running   0              55d
es            elasticsearch-data-2                              1/1     Running   0              21d
kube-system   calico-node-n62mj                                 1/1     Running   32 (76d ago)   155d
kube-system   coredns-69766b66b-2zlsr                           1/1     Running   0              69d
kube-system   kube-proxy-jprg2                                  1/1     Running   1              155d
kube-system   local-px-int-njs5q                                1/1     Running   5 (55m ago)    70d
kube-system   nginx-proxy-ip-10-13-193-24.pwx.purestorage.com   1/1     Running   1 (76d ago)    155d
kube-system   nodelocaldns-7pm2x                                1/1     Running   1              155d
kube-system   portworx-api-kks4x                                2/2     Running   0              70d
kube-system   portworx-kvdb-cjjpx                               1/1     Running   6 (55m ago)    155d
kube-system   px-csi-ext-6c89d5d4f4-7d6fs                       4/4     Running   0              34d
kube-system   stork-996664894-2xdf9                             1/1     Running   0              4m45s
kube-system   stork-scheduler-7d5579975-5zrdc                   1/1     Running   0              34d
mysql         mysql-6d9f9d7947-b9t9l                            1/1     Running   0              68d
mysql5        mysql-6b89f6b6d4-qlxv8                            1/1     Running   0              55d
mysql8        mysql-bb9686cdb-qz7mh                             1/1     Running   0              68d
```

After stopping px on that node, the drivermonitor picked up and started deleting the pods
```
➜  ~ ks logs stork-996664894-6bwnq | grep -i "Deleting pod from node"
time="2024-08-21T11:32:25Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-16 PodName=busybox-deployment-6dcd7756f9-sqdnm
time="2024-08-21T11:32:25Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-6 PodName=busybox-deployment-6dcd7756f9-z2nnz
time="2024-08-21T11:32:25Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=es PodName=elasticsearch-data-0
time="2024-08-21T11:32:25Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=mysql PodName=mysql-6d9f9d7947-b9t9l
time="2024-08-21T11:32:25Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=mysql5 PodName=mysql-6b89f6b6d4-qlxv8
time="2024-08-21T11:32:27Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-10 PodName=busybox-deployment-6dcd7756f9-7q5r6
time="2024-08-21T11:32:27Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-4 PodName=busybox-deployment-6dcd7756f9-zqrm7
time="2024-08-21T11:32:27Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=es PodName=elasticsearch-data-2
time="2024-08-21T11:32:27Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=mysql8 PodName=mysql-bb9686cdb-qz7mh
time="2024-08-21T11:32:27Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-2 PodName=busybox-deployment-6dcd7756f9-w7v7r
time="2024-08-21T11:32:30Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-14 PodName=busybox-deployment-6dcd7756f9-ghz9f
time="2024-08-21T11:32:30Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-20 PodName=busybox-deployment-654cf6f86f-zf9r2
time="2024-08-21T11:32:30Z" level=info msg="Deleting Pod from Node ip-10-13-193-24.pwx.purestorage.com due to volume driver status: Offline (STATUS_OFFLINE)" Namespace=busy-7 PodName=busybox-deployment-6dcd7756f9-bkwt9

```
After pods using px volumes got evicted from the offline node 
```
➜  stork git:(PWX-38468) ✗ kubectl get pods --all-namespaces --field-selector spec.nodeName=ip-10-13-193-24.pwx.purestorage.com
NAMESPACE     NAME                                              READY   STATUS    RESTARTS       AGE
kube-system   calico-node-n62mj                                 1/1     Running   32 (76d ago)   155d
kube-system   coredns-69766b66b-2zlsr                           1/1     Running   0              69d
kube-system   kube-proxy-jprg2                                  1/1     Running   1              155d
kube-system   local-px-int-njs5q                                0/1     Running   6 (2m4s ago)   70d
kube-system   nginx-proxy-ip-10-13-193-24.pwx.purestorage.com   1/1     Running   1 (76d ago)    155d
kube-system   nodelocaldns-7pm2x                                1/1     Running   1              155d
kube-system   portworx-api-kks4x                                1/2     Running   0              70d
kube-system   portworx-kvdb-cjjpx                               0/1     Running   7 (101s ago)   155d
kube-system   stork-996664894-2xdf9                             1/1     Running   0              17m
kube-system   stork-scheduler-7d5579975-5zrdc                   1/1     Running   0              34d
```

**UT**:
```
➜  stork git:(PWX-38468) ✗ /usr/local/go/bin/go test -timeout 1200s -tags unittest,integrationtest  github.com/libopenstorage/stork/pkg/monitor
ok  	github.com/libopenstorage/stork/pkg/monitor	663.688s

Removing testVolumeAttachmentCleanup UT as it is not working.
      1. With existing mocking, the function can only give a fixed output but need dynamic output for InspectNode output
      2. With MockDriver, GetPodsByNode is not giving the desired output as the fieldselector does not work with mockdriver

```

### CODE REVIEW

**Review by strivedi-px — CHANGES_REQUESTED** [review]

**Review by strivedi-px — APPROVED** [review]

File: pkg/monitor/monitor.go [hunk 1]
```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
 		log.Errorf("Error cleaning up volume attachments: %v", err)
 	}
```
Comment (pp511) [inline pkg/monitor/monitor.go:299]: Some optimisations can be done inside cleanupVolumeAttachmentsByNode().
vaList, err := storage.Instance().ListVolumeAttachments() is being called every time. There is scope of calling this once and use it for the current iteration of the driver monitor.

Comment (pp511) [reply pkg/monitor/monitor.go:299]: Also isSameNode() is being called inside the same function which calls core.Instance().GetNodeByName() which is an unnecessary API call. We have a K8s node name list in the map already. We could either create a map of name to K8s node object and use it. Or just iterate over it every time. This would avoid making an extra api call. Let me know what you think.

File: pkg/monitor/monitor.go [hunk 2]
```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
...
+
+	// Do batch deletes of rest of the pods
+	if len(podsSelectedForDeletion) > 0 {
+		m.batchDeleteOfPodsFromOfflineNodes(podsSelectedForDeletion, node, false)
+	}
+}
+
+// batchDeleteOfPodsFromOfflineNodes deletes the pods in batches of batchSize
+// and increments the HealthCounter by the number of pods deleted
+func (m *Monitor) batchDeleteOfPodsFromOfflineNodes(
```
[the remark this answers is not in the document]
Comment (pp511) [reply pkg/monitor/monitor.go:333]: +1

**Review by pp511 — COMMENTED** [review]
Overall looks ok. Sharing some suggestions...

File: pkg/monitor/monitor.go [hunk 3]
```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
...
-			if err != nil {
-				storklog.PodLog(&pod).Errorf("Error deleting pod: %v", err)
-				continue
-			}
-			HealthCounter.Inc()
 		}
+
+		err := core.Instance().DeletePods(pods[i:end], true)
+		if err != nil {
+			log.Errorf("Error deleting pods: %v", err)
```
Comment (pp511) [inline pkg/monitor/monitor.go:368]: It is a known problem that PX can come online and we may continue to delete the Pods. Does it not make sense to check the status of PX after each batch?

File: pkg/monitor/monitor.go [hunk 4]
```diff
@@ -396,3 +458,25 @@ func (m *Monitor) cleanupVolumeAttachmentsByNode(node *volume.NodeInfo) error {
...
 	return nil
 }
+
+func (m *Monitor) getVolumeDriverNodesToK8sNodeMap(driverNodes []*volume.NodeInfo) map[string]*v1.Node {
+	nodeMap := make(map[string]*v1.Node)
+
+	k8sNodes, err := core.Instance().GetNodes()
+	if err != nil {
+		log.Errorf("Error getting nodes in monitor: %v", err)
+		return nodeMap
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:468]: It's not an infinite loop though. For every cycle of monitoring, it will go over the nodes and only if the node status is offline, then we are trying to put that in the map. In one cycle of drivermonitor (default value of 2mins) , the k8s call happens one time to create the map (for successful case) but if k8s itself does not response, then that call can happen for offline number of nodes. 

Inside getVolumeDriverNodesToK8sNodeMap , there is no API calls to PX.

[the remark this answers is not in the document]
Comment (pp511) [reply pkg/monitor/monitor.go:468]: Ok. I see a sleep after the inner for loop. It should be ok then.

File: pkg/monitor/monitor.go [hunk 5]
```diff
@@ -118,6 +119,13 @@ func (m *Monitor) isSameNode(k8sNodeName string, driverNode *volume.NodeInfo) bo
 	return volume.IsNodeMatch(node, driverNode)
 }
 
+func (m *Monitor) isSameK8sNode(k8sNode *v1.Node, driverNode *volume.NodeInfo) bool {
+	if k8sNode.Name == driverNode.Hostname {
+		return true
+	}
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:125]: Other packages like extender is using volume.IsNodeMatch, so kept this check outside.

File: pkg/monitor/monitor.go [hunk 6]
```diff
@@ -245,14 +263,14 @@ func (m *Monitor) driverMonitor() {
 	}
 }
 
-func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
+func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo, k8sNode *v1.Node) {
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: Done

File: pkg/monitor/monitor.go [hunk 7]
```diff
@@ -228,9 +237,18 @@ func (m *Monitor) driverMonitor() {
 				// For any Running pod on that node using volume by the driver, kill the pod
 				// Degraded nodes are not considered offline and pods are not deleted from them.
 				if node.Status == volume.NodeOffline || node.Status == volume.NodeDegraded {
+					// Only initialize the map when it is absolutely necessary
+					if len(driverNodeToK8sNodeMap) == 0 {
+						driverNodeToK8sNodeMap = m.getVolumeDriverNodesToK8sNodeMap(nodes)
+					}
+					if _, ok := driverNodeToK8sNodeMap[node.StorageID]; !ok {
+						log.Errorf("Node with scheduler Name %v and storage ID %v is not found in k8s cluster", node.SchedulerID, node.StorageID)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: Done

File: pkg/monitor/monitor.go [hunk 8]
```diff
@@ -228,9 +237,18 @@ func (m *Monitor) driverMonitor() {
 				// For any Running pod on that node using volume by the driver, kill the pod
 				// Degraded nodes are not considered offline and pods are not deleted from them.
 				if node.Status == volume.NodeOffline || node.Status == volume.NodeDegraded {
+					// Only initialize the map when it is absolutely necessary
+					if len(driverNodeToK8sNodeMap) == 0 {
+						driverNodeToK8sNodeMap = m.getVolumeDriverNodesToK8sNodeMap(nodes)
+					}
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:244]: Then in every cycle of 2mins we have to reconstruct it even if there are no offline nodes.
Also we can't take it outside of the outer for loop, since the data can become stale.

Took the approach of only initialising it if there is an offline node in every cycle.

File: pkg/monitor/monitor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: Done

File: pkg/monitor/monitor.go [hunk 9]
```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
...
+	node *volume.NodeInfo,
+	csiPods bool) {
+
+	// Do a fresh node status check before deleting pods
+	n, err := m.Driver.InspectNode(node.StorageID)
+	if err != nil {
+		log.Errorf("Error inspecting node %v in batch deletion of pods: %v", node.StorageID, err)
+		return
+	}
+	if !m.isNodeOffline(n) {
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: Good catch. We want to still return if the is not offline but here changed the log line.

File: pkg/monitor/monitor.go [hunk 10]
```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
...
+		log.Infof("Volume driver on node %v (%v) is still offline (%v) in batch deletion of pods", node.Hostname, node.StorageID, n.RawStatus)
+		return
+	}
+	log.Infof("Going to delete %d number of pods from node %v in batch", len(pods), node.Hostname)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: done

File: pkg/monitor/monitor.go (hunk 3 above)
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: Good suggestion, now moved the Inspectnode into each batch, also added a default 30s interval between batches of pod deletes.

File: pkg/monitor/monitor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:473]: As explained in the below comment, we dont want to return if the call fails intermittently wrt some node in the node looping. The idea is to try to act on any offline node in the cycle if there are any. Now we are going to limit the number of goroutines for cleanupPods wrt node too.

File: pkg/monitor/monitor.go [hunk 11]
```diff
@@ -280,32 +298,76 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
...
+	csiPods bool) {
+
+	// Do a fresh node status check before deleting pods
+	n, err := m.Driver.InspectNode(node.StorageID)
+	if err != nil {
+		log.Errorf("Error inspecting node %v in batch deletion of pods: %v", node.StorageID, err)
+		return
+	}
+	if !m.isNodeOffline(n) {
+		log.Infof("Volume driver on node %v (%v) is still offline (%v) in batch deletion of pods", node.Hostname, node.StorageID, n.RawStatus)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go]: modified.

File: pkg/monitor/monitor.go (hunk 1 above)
Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:297]: As discussed not addressing the first comment of doing `ListVolumeAttachments` once as we are going to limit the number of goroutines.

Number of GetNodeByName is replaced with doing one GetNode call and create the map and use it for one cycle of handling offline nodes in drivermonitor.

**Comment by diptiranjanpx** [comment]
Integration test- https://jenkins.pwx.dev.purestorage.com/job/Users/job/strivedi/job/dup-healthmonitor/2/

**Review by strivedi-px — APPROVED** [review]

**Review by strivedi-px — APPROVED** [review]

File: pkg/monitor/monitor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pp511) [reply pkg/monitor/monitor.go]: Do you think moving the condition inside isSameK8sNode will break something?

File: pkg/monitor/monitor.go [hunk 12]
```diff
@@ -228,9 +227,26 @@ func (m *Monitor) driverMonitor() {
...
+					}
+					if _, ok := driverNodeToK8sNodeMap[node.StorageID]; !ok {
+						log.Errorf("Node with scheduler name %v and storage ID %v is not found in k8s cluster", node.SchedulerID, node.StorageID)
+						continue
+					}
+					k8sNode := driverNodeToK8sNodeMap[node.StorageID]
 					m.wg.Add(1)
 					// wait for 1 min if node is upgrading
-					go m.cleanupDriverNodePods(node)
+					go m.cleanupDriverNodePods(node, k8sNode, k8sNodeNameToNodeMap)
```
Comment (pp511) [inline pkg/monitor/monitor.go:249]: I didn't understand the comment // just before this line. "// wait for 1 min if node is upgrading"

Comment (pp511) [inline pkg/monitor/monitor.go:249]: I remember you mentioning that we are going to limit the number of go routines using the sched package. Is it being done in a separate PR?

Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:249]: Yes, that is being handled as part of another ticket.

Comment (diptiranjanpx) [reply pkg/monitor/monitor.go:249]: Will remove this line.
